### PR TITLE
[slspath] make paths generic so the formula can be placed anywhere

### PR DESCRIPTION
**Pull request by ChrisLundquist**
At the moment we hardcode the slspath to `nginx` in some places.
This prevents me from placing many `saltstack-formulas` together, and forces me to add a repo root for each one.
Moreover, if I had two different nginx formulas, I wouldn't be able to rename them to resolve the collision, (or I would get unexpected behavior if it did work by loading the other formula)

Basically, it allows my top.sls to look like this:

``` yaml
"web*":
  - saltstack.nginx.nginx
  - saltstack.ntp.ntp
  - my.nginx.vhost
  - my.nginx.tweaks
  - my.ntp.servers
```

The `ntp.ntp` is a little ugly, but that's how it works out with the formula repo structure.
This approach allows me to use the community formulas for the broad strokes, and make any specific changes that no one else would be interested in with my own formulas.

### CODE REVIEW

**Comment by gravyboat** [comment]
This one is definitely going to require some discussion. The point of the formulas is to provide users with a good starting place they can branch out from. The usage of a custom sls path that isn't as clearly defined introduces additional complexity which isn't beneficial for new users and is easily rectified via a fork and modification like you've done. I'm going to tag @whiteinge to get his opinion on this PR.

**Comment by ChrisLundquist** [comment]
> The point of the formulas is to provide users with a good starting place they can branch out from

I'm definitely on board there. When learning a new configuration tool one of the first things I like to do is look at the community offerings to absorb the best practices and idioms of the language / tool.

I *also* think that using the community offerings is important, and making it as easy as possible is important. This repo as a excellent history of responding to issues and merging pull requests, and it isn't always possible to merge things. Eventually, this lead to huge fragmentation in the chef world, spawning  tools like chef librarian and berkshelf to be written to manage external dependencies.

In the above chef case, the one thing it had going for it was that you could rename cookbooks to things like "heavywater-redis" and it would still work. Currently in this salt repo, renaming the formula breaks it. Having two different teams or products with slightly different needs using the same salt repo doesn't seem like a far fetched case to me. It is in fact this case that brought me here. I'm trying to merge my salt repo with another team and I want to use this formula.

I took a brief look at Salt Package Manager to see if it was like `bundler`, `npm`, or `berkshelf`, and it didn't seem to. It seems to zip repos and unzip them in case you can't or won't use the git backend.

To recap, to me this is really about formula renaming. I want to name this `saltstack.nginx` and I can't.
Maintaining my fork is very costly as each path that makes it into this repo would have to be updated. In reality I would diverge and cause community fragmentation.

**Comment by ChrisLundquist** [comment]
I've still been thinking about this and tried one more approach; symlinking the formulas into my repo. That didn't quite work either unless they are in top level `nginx` folder for example. It does the remove the ugly `nginx.nginx` though. It seems the only way to use the formula is via `file_roots` in `/etc/salt/master` and there is no way to avoid naming conflicts with this package short of this PR. It is a little counter intuitive that I must change my master to use a formula, rather than the formula adapt to the master. (With added pain that I can't signal the master to reload configuration, only configure a refresh interval or restart)

Renaming is important to me because some of my infrastructure uses Salt + Docker and other parts use Salt to manage long lived VMs / bare metal. So, there is a very different formula to manage the nginx docker container compared to my usage of this formula to manage nginx on a long lived vm.

**Comment by whiteinge** [comment]
Sorry for the slow reply.

This is an interesting use-case. Painting with a (very) broad brush: The vast majority use for most services is to have a single server instance per machine. For services that deviate from that and are more commonly installed in multi-tenant or multi-head configurations the formula should be written to accommodate those via looping and Pillar configuration. Two real-world examples, respectively are: the Apache/Nginx methods of configuring separate vhosts by placing separate .conf files; and the Debian-style configuration of memcached where the init script is written to invoke multiple instances of the server, one for each found .conf file. However, all that said, more flexibility is always a good thing and you don't always have control over everything. Case in point, the CentOS memcached init script cannot invoke multiple server instances out-of-box; that's a Debian-ism. Read on...

@gravyboat is right that this pattern applied to all formulas presents a lot of visual noise in addition to logistical troubles such as convention enforcement across all the formula repos. Currently formulas are pretty single-mindedly organized as drop-ins to the `file_roots` or `gitfs_remots` arrays. Spitballing: I think this would be better handled as a Salt addition -- some kind of aliasing syntax, e.g.: `gitfs_remotes: [{'nginx-two': 'https://github.com/saltstack-formulas/nginx-formula.git'}]` that can rewrite slspath on the fly when imported. @ChrisLundquist If that sounds like a good apprach to you please open an issue in the Salt repo, cross-link this issue, and feel free to tag me there. I think it would be a nice Salt addition.

**Comment by ChrisLundquist** [comment]
> This is an interesting use-case. Painting with a (very) broad brush: The vast majority use for most services is to have a single server instance per machine.

You're picking up on another one of my stumbling points, parameterized states. Currently, the only cited methods are looping over pillar variables or complex jinja macros. [The stateconf rendered](https://docs.saltstack.com/en/latest/ref/renderers/all/salt.renderers.stateconf.html) is the only hint of this idea.

The real killer feature here is renaming / moving. Think of it like being able to install a foreign debian package into `/opt` or `/usr/local`

> Spitballing: I think this would be better handled as a Salt addition -- some kind of aliasing syntax, e.g.: gitfs_remotes: [{'nginx-two': 'https://github.com/saltstack-formulas/nginx-formula.git'}] that can rewrite slspath on the fly when imported.

You undoubtably know saltstack far better than me, though it seems like in the feature you described, the hard coded path would still cause issues if the formula isn't using `slspath` for imports and sources.
 
Making a shim in `top.sls` is another aproach, but it doesn't seem like it would work. It seems you'd be in the same spot though because if you made `nginx-three` it would also be translated to `nginx` causing the same conflict. Naturally, each formula wouldn't know about the other, so if they did things like `import nginx/map.jina` unexpected behavior might happen.

It seems like each formula would need to be taught to be re-rooted in either case.

Thank you for your perspective! I'll open the issue on salt.

**Comment by ChrisLundquist** [comment]
It is worth mentioning that one widely used solution is relative paths. Currently, all paths in salt seem to be absolute and rooted in the salt root (which makes sense). If paths were implicitly relative to the formula like shells and `pwd`, you could achieve this as `{{ slspath }}` would be implied everywhere it is in this PR.

The above is a pretty large change though and could easily be a breaking change unless you fell back to the current behavior. The downside of supporting both is it might cause some bugs.

**Comment by noelmcloughlin** [comment]
See https://github.com/saltstack/salt/issues/30857
